rpc/rest/security/rest_security.go

```diff
@@ -0,0 +1,304 @@
+		if authHeader == "" {
+			a.formatter.Text(w, http.StatusUnauthorized, "401 Unauthorized: authorization header required")
+			return
+		}
+		bearerToken := strings.Split(authHeader, " ")
+		if len(bearerToken) != 2 {
+			a.formatter.Text(w, http.StatusUnauthorized, "401 Unauthorized: invalid authorization token")
+			return
+		}
+		token, err := jwt.Parse(bearerToken[1], func(token *jwt.Token) (interface{}, error) {
+			if _, ok := token.Method.(*jwt.SigningMethodHMAC); !ok {
+				return nil, fmt.Errorf("error parsing token")
+			}
+			return []byte(secret), nil
+		})
+		if err != nil {
+			errStr := fmt.Sprintf("500 internal server error: %s", err)
+			a.formatter.Text(w, http.StatusInternalServerError, errStr)
+			return
+		}
+
+		if err := a.validateToken(token.Raw, req.URL.Path); err != nil {
+			errStr := fmt.Sprintf("401 Unauthorized: %v", err)
+			a.formatter.Text(w, http.StatusUnauthorized, errStr)
+			return
+		}
+
+		provider.ServeHTTP(w, req)
+	})
+}
+
+// Register authenticator-wide security handlers
+func (a *authenticator) registerSecurityHandlers() {
+	a.router.HandleFunc(login, a.createTokenEndpoint).Methods(http.MethodPost)
+	a.router.HandleFunc(logout, a.invalidateTokenEndpoint).Methods(http.MethodPost)
+}
+
+// Validates credentials and provides new token
+func (a *authenticator) createTokenEndpoint(w http.ResponseWriter, req *http.Request) {
+	name, errCode, err := a.validateCredentials(req)
+	if err != nil {
+		a.formatter.Text(w, errCode, err.Error())
+		return
+	}
+	var claims jwt.MapClaims = make(map[string]interface{})
+	if a.expTime != 0 {
+		claims["exp"] = time.Now().Add(time.Minute * time.Duration(a.expTime)).Unix()
+	}
+	token := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)
+	tokenString, err := token.SignedString([]byte(secret))
+	if err != nil {
+		errStr := fmt.Sprintf("500 internal server error: faield to sign token: %v", err)
+		a.formatter.Text(w, http.StatusInternalServerError, errStr)
+		return
+	}
+	a.tokenDb[name] = tokenString
+	a.formatter.Text(w, http.StatusOK, tokenString)
+	return
+}
+
+// Removes token endpoint from the DB. During processing, token will not be found and will be considered as invalid.
+func (a *authenticator) invalidateTokenEndpoint(w http.ResponseWriter, req *http.Request) {
+	name, _, ok := req.BasicAuth()
+	if !ok {
+		a.formatter.Text(w, http.StatusInternalServerError, "500 internal server error: unable to process authentication data")
+		return
+	}
+	delete(a.tokenDb, name)
+}
+
+// Validates credentials, returns name and error code/message if invalid
+func (a *authenticator) validateCredentials(req *http.Request) (string, int, error) {
+	name, pass, ok := req.BasicAuth()
+	if !ok {
+		return name, http.StatusInternalServerError, errors.Errorf("500 internal server error: unable to process authentication data")
+	}
+	user := a.getUser(name)
+	if user == nil {
+		return name, http.StatusUnauthorized, errors.Errorf("401 unauthorized: user name does not exist")
```

Comment:
For security reasons, the error message should be exactly same for invalid user and invalid password. Otherwise, an attacker will be able to do brute force attack more easily.

Reply:
I actually hate this approach, because when I try to remember some credentials, I never know what I put wrong :-)

Reply:
Double-edged sword I know..